src/class/dfu/dfu_rt_device.c

```diff
@@ -106,34 +117,67 @@ bool dfu_rtd_control_xfer_cb(uint8_t rhport, uint8_t stage, tusb_control_request
   // Handle class request only from here
   TU_VERIFY(request->bmRequestType_bit.type == TUSB_REQ_TYPE_CLASS);
 
-  switch ( request->bRequest )
+  TU_LOG2("  DFU Request: %s\r\n", tu_lookup_find(&_dfu_request_table, request->bRequest));
+  switch (request->bRequest)
   {
     case DFU_REQUEST_DETACH:
-      tud_control_status(rhport, request);
-      tud_dfu_runtime_reboot_to_dfu_cb();
+    {
+      if (_dfu_state_ctx.state == APP_IDLE)
+      {
+        _dfu_state_ctx.state = APP_DETACH;
+        if ((_dfu_state_ctx.attrs & DFU_FUNC_ATTR_WILL_DETACH_BITMASK) == 1)
+        {
+          tud_dfu_runtime_reboot_to_dfu_cb();
+        } else {
+          tud_dfu_runtime_detach_start_timer_cb(request->wValue);
+        }
+      } else {
+        TU_LOG2("  DFU Unexpected request during state %s: %u\r\n", tu_lookup_find(&_dfu_mode_state_table, _dfu_state_ctx.state), request->bRequest);
+        return false;
+      }
+    }
     break;
 
     case DFU_REQUEST_GETSTATUS:
     {
-      // status = OK, poll timeout = 0, state = app idle, istring = 0
-      uint8_t status_response[6] = { 0, 0, 0, 0, 0, 0 };
-      tud_control_xfer(rhport, request, status_response, sizeof(status_response));
+      dfu_status_req_payload_t resp;
+      resp.bStatus = _dfu_state_ctx.status;
+      memset((uint8_t *)&resp.bwPollTimeout, 0x00, 3);  // Value is ignored
+      resp.bState = _dfu_state_ctx.state;
+      resp.iString = ( tud_dfu_runtime_get_status_desc_table_index_cb ) ? tud_dfu_runtime_get_status_desc_table_index_cb() : 0;
+
+      tud_control_xfer(rhport, request, &resp, sizeof(dfu_status_req_payload_t));
     }
     break;
 
-    default: return false; // stall unsupported request
+    case DFU_REQUEST_GETSTATE:
+    {
+      tud_control_xfer(rhport, request, &_dfu_state_ctx.state, 1);
+    }
+    break;
+
+    default:
+    {
+      TU_LOG2("  DFU Nonstandard Runtime Request: %u\r\n", request->bRequest);
+      return ( tud_dfu_runtime_req_nonstandard_cb ) ? tud_dfu_runtime_req_nonstandard_cb(rhport, stage, request) : false;
```

Comment:
please removed, we don't support vendor request here. Vendor need to implement their own if needed. Here is how raspberrypi add their own. https://github.com/raspberrypi/pico-sdk/pull/197

Reply:
Resolved in 2e2dc7b